File: python_plots_generator/Server/DTOS/LinearRegressionDTO.py
from flask_restx import fields, Namespace

from Config.constants import FORMATTERS_LIST

"""

    {
        "src": {
            "main": {
                "path": "spark/temp/evol/cadiz/evol",
                "axis": {
                    "x": {
                        "name": "date",
                        "format": "timestamp"
                    },
                    "y": {
                        "name": "temp_daily_avg"
                    }
                }
            },
            "regression": {
                "path": "spark/temp/evol/cadiz/evol",
                "names": {
                    "slope": "beta_1",
                    "intercept": "beta_0"
                }
            }
        },
        "dest" : {
            "path": "temp/evol/cadiz/evol",
            "filename": "plot_test",
            "export_png": true
        },
        "style": {
            "lettering": {
                "title": "TITLE TEST",
                "subtitle": "SUBTITLE TEST",
                "x_label": "Date",
                "y_1_label": "Var (u)",
                "y_2_label": "Var (u)"
            },
            "figure_1": {
                "name": "var",
                "color": "#4169e1"
            },
            "figure_2": {
                "name": "var",
                "color": "#4169e1"
            },
            "margin": {
                "left": 120,
                "right": 120,
                "top": 100,
                "bottom": 100
            },
            "legend": {
                "y_offset": -0.05
            }
        }
    }


"""

class LinearRegressionDTO:
    def __init__(self, ns: Namespace):
        self.post_input = _create_input_post_dto(ns)


def _create_input_post_dto(ns: Namespace):
    return ns.model('LinearRegressionInput', {
        'src': fields.Nested(ns.model('LinearRegressionSrc', {
            'main': fields.Nested(ns.model('LinearRegressionSrcMain', {
                'path': fields.String(required=True, description="Relative route to main data"),
                'axis': fields.Nested(ns.model('LinearRegressionSrcAxis', {
                    'x': fields.Nested(ns.model('LinearRegressionSrcAxisX', {
                        'name': fields.String(reqired=True, description="X Column name"),
                        'format': fields.String(reqired=False, description="Column format", enum=FORMATTERS_LIST),
                    }), required=True),
                    'y': fields.Nested(ns.model('LinearRegressionSrcAxisY', {
                        'name': fields.String(reqired=True, description="Y Column name")
                    }), required=True)
                }), required=True)
            }), required=True),
            'regression': fields.Nested(ns.model('LinearRegressionSrcRegression', {
                'path': fields.String(required=True, description="Relative route to regression data"),
                'names': fields.Nested(ns.model('LinearRegressionSrcRegressionNames', {
                    'slope': fields.String(required=True, description="Slope Column name"),
                    'intercept': fields.String(required=True, description="Intercept Column name"),
                }), required=True)
            }), required=True),
        })),
        'dest': fields.Nested(ns.model('LinearRegressionDest', {
            'path': fields.String(required=True, description="Relative output route"),
            'filename': fields.String(required=True, description="Filename of the chart"),
            'export_png': fields.Boolean(required=True, description="Export png"),
        }), required=True),
        'style': fields.Nested(ns.model('LinearRegressionStyle', {
            'lettering': fields.Nested(ns.model('LinearRegressionStyleLettering', {
                'title': fields.String(required=True, description="Title of the chart"),
                'subtitle': fields.String(required=False, description="Subtitle of the chart"),
                'x_label': fields.String(required=True, description="X axis label"),
                'y_1_label': fields.String(required=True, description="Y 1 axis label"),
                'y_2_label': fields.String(required=True, description="Y 2 axis label"),
            }), required=True),
            'figure_1': fields.Nested(ns.model('LinearRegressionStyleFigure1', {
                'name': fields.String(required=True, description="Figure 1 name"),
                'color': fields.String(required=True, description="Figure 1 color"),
            }), required=True),
            'figure_2': fields.Nested(ns.model('LinearRegressionStyleFigure2', {
                'name': fields.String(required=True, description="Figure 2 name"),
                'color': fields.String(required=True, description="Figure 2 color"),
            }), required=True),
            'margin': fields.Nested(ns.model('LinearRegressionStyleMargin', {
                'left': fields.Float(required=True, description="Margin left"),
                'right': fields.Float(required=True, description="Margin right"),
                'top': fields.Float(required=True, description="Margin top"),
                'bottom': fields.Float(required=True, description="Margin bottom")
            }), required=False),
            'legend': fields.Nested(ns.model('LinearRegressionStyleLegend', {
                'y_offset': fields.Float(required=True, description="Legend Y axis offset"),
            }), required=False),
        }), required=True)
    })

File: scala_big_data_core/src/main/scala/Config/PlotGenerationConf/Execution/DTO/LinearRegressionDTO.scala
package Config.PlotGenerationConf.Execution.DTO

import upickle.default.{ReadWriter, macroRW}
import upickle.implicits.key

case class LinearRegressionDTOAxisComponent(
  @key("name") name: String,
  @key("format") format: Option[String]
)
object LinearRegressionDTOAxisComponent { implicit val rw: ReadWriter[LinearRegressionDTOAxisComponent] = macroRW }

case class LinearRegressionDTOAxis(
  @key("x") x: LinearRegressionDTOAxisComponent,
  @key("y") y: LinearRegressionDTOAxisComponent
)
object LinearRegressionDTOAxis { implicit val rw: ReadWriter[LinearRegressionDTOAxis] = macroRW }

case class LinearRegressionDTOSrcMainComponent(
  @key("path") path: String,
  @key("axis") axis: LinearRegressionDTOAxis
)
object LinearRegressionDTOSrcMainComponent { implicit val rw: ReadWriter[LinearRegressionDTOSrcMainComponent] = macroRW }

case class LinearRegressionDTOSrcRegressionComponentNamesComponent(
  @key("slope") slope: String,
  @key("intercept") intercept: String
)
object LinearRegressionDTOSrcRegressionComponentNamesComponent { implicit val rw: ReadWriter[LinearRegressionDTOSrcRegressionComponentNamesComponent] = macroRW }

case class LinearRegressionDTOSrcRegressionComponent(
  @key("path") path: String,
  @key("names") names: LinearRegressionDTOSrcRegressionComponentNamesComponent
)
object LinearRegressionDTOSrcRegressionComponent { implicit val rw: ReadWriter[LinearRegressionDTOSrcRegressionComponent] = macroRW }

case class LinearRegressionDTOSrc(
  @key("main") main: LinearRegressionDTOSrcMainComponent,
  @key("regression") regression: LinearRegressionDTOSrcRegressionComponent
)
object LinearRegressionDTOSrc { implicit val rw: ReadWriter[LinearRegressionDTOSrc] = macroRW }

case class LinearRegressionDTODest(
  @key("path") path: String,
  @key("filename") filename: String,
  @key("export_png") exportPng: Boolean
)
object LinearRegressionDTODest { implicit val rw: ReadWriter[LinearRegressionDTODest] = macroRW }

case class LinearRegressionDTOLettering(
  @key("title") title: String,
  @key("subtitle") subtitle: String,
  @key("x_label") xLabel: String,
  @key("y_1_label") y1Label: String,
  @key("y_2_label") y2Label: String
)
object LinearRegressionDTOLettering { implicit val rw: ReadWriter[LinearRegressionDTOLettering] = macroRW }

case class LinearRegressionDTOFigure(
  @key("name") name: String,
  @key("color") color: String
)
object LinearRegressionDTOFigure { implicit val rw: ReadWriter[LinearRegressionDTOFigure] = macroRW }

case class LinearRegressionDTOMargin(
  @key("left") left: Float,
  @key("right") right: Float,
  @key("top") top: Float,
  @key("bottom") bottom: Float
)
object LinearRegressionDTOMargin { implicit val rw: ReadWriter[LinearRegressionDTOMargin] = macroRW }

case class LinearRegressionDTOLegend(
  @key("y_offset") yOffset: Float
)
object LinearRegressionDTOLegend { implicit val rw: ReadWriter[LinearRegressionDTOLegend] = macroRW }

case class LinearRegressionDTOStyle(
  @key("lettering") lettering: LinearRegressionDTOLettering,
  @key("figure_1") figure1: LinearRegressionDTOFigure,
  @key("figure_2") figure2: LinearRegressionDTOFigure,
  @key("margin") margin: Option[LinearRegressionDTOMargin],
  @key("legend") legend: Option[LinearRegressionDTOLegend]
)
object LinearRegressionDTOStyle { implicit val rw: ReadWriter[LinearRegressionDTOStyle] = macroRW }

case class LinearRegressionDTO (
  @key("src") src: LinearRegressionDTOSrc, 
  @key("dest") dest: LinearRegressionDTODest, 
  @key("style") style: LinearRegressionDTOStyle
)
object LinearRegressionDTO { implicit val rw: ReadWriter[LinearRegressionDTO] = macroRW }
